File: android/app/src/main/java/ai/moataz/app/ServerUrlPolicy.java
package ai.moataz.app;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Locale;

final class ServerUrlPolicy {
    private ServerUrlPolicy() {}

    static String normalize(String raw) {
        if (raw == null) throw new IllegalArgumentException("Server URL is required");
        String value = raw.trim();
        if (value.isEmpty()) throw new IllegalArgumentException("Server URL is required");
        if (!value.contains("://")) value = "http://" + value;

        try {
            URI uri = new URI(value);
            String scheme = uri.getScheme() == null ? "" : uri.getScheme().toLowerCase(Locale.ROOT);
            if (!scheme.equals("http") && !scheme.equals("https")) {
                throw new IllegalArgumentException("Only http:// and https:// are supported");
            }
            if (uri.getHost() == null || uri.getHost().isBlank()) {
                throw new IllegalArgumentException("Server URL must include a host");
            }
            if (uri.getUserInfo() != null) {
                throw new IllegalArgumentException("Credentials are not allowed in the server URL");
            }
            URI normalized = new URI(
                scheme,
                null,
                uri.getHost().toLowerCase(Locale.ROOT),
                uri.getPort(),
                cleanPath(uri.getPath()),
                null,
                null
            );
            String result = normalized.toString();
            return result.endsWith("/") ? result.substring(0, result.length() - 1) : result;
        } catch (URISyntaxException | IllegalArgumentException error) {
            if (error instanceof IllegalArgumentException) throw (IllegalArgumentException) error;
            throw new IllegalArgumentException("Invalid server URL", error);
        }
    }

    static boolean isTrustedOrigin(String serverUrl, String trustedOrigin) {
        if (trustedOrigin == null || trustedOrigin.isBlank()) return false;
        try {
            URI server = new URI(normalize(serverUrl));
            URI trusted = new URI(normalize(trustedOrigin));
            return originPort(server) == originPort(trusted)
                && server.getScheme().equalsIgnoreCase(trusted.getScheme())
                && server.getHost().equalsIgnoreCase(trusted.getHost());
        } catch (RuntimeException | URISyntaxException ignored) {
            return false;
        }
    }

    private static int originPort(URI uri) {
        if (uri.getPort() >= 0) return uri.getPort();
        return "https".equalsIgnoreCase(uri.getScheme()) ? 443 : 80;
    }

    private static String cleanPath(String path) {
        if (path == null || path.isBlank() || "/".equals(path)) return "";
        String cleaned = path;
        while (cleaned.endsWith("/")) cleaned = cleaned.substring(0, cleaned.length() - 1);
        return cleaned;
    }
}

File: android/app/src/main/java/ai/moataz/app/ServerUrlPolicy.kt
package ai.moataz.app

import java.net.URI
import java.util.Locale

/**
 * Normalizes the user-selected moataz ai Gateway origin.
 *
 * The Android client talks to DeerFlow/moataz ai APIs directly. It never
 * accepts credentials embedded in the URL and it strips path/query fragments
 * so all API routes are resolved from one stable origin.
 */
object ServerUrlPolicy {
    @JvmStatic
    fun normalize(raw: String?): String {
        require(!raw.isNullOrBlank()) { "Server URL is required" }
        val candidate = raw.trim().let { if ("://" in it) it else "http://$it" }
        val uri = runCatching { URI(candidate) }
            .getOrElse { throw IllegalArgumentException("Invalid server URL", it) }
        val scheme = uri.scheme?.lowercase(Locale.ROOT).orEmpty()
        require(scheme == "http" || scheme == "https") { "Only http:// and https:// are supported" }
        require(!uri.host.isNullOrBlank()) { "Server URL must include a host" }
        require(uri.userInfo == null) { "Credentials are not allowed in the server URL" }

        val normalized = URI(
            scheme,
            null,
            uri.host.lowercase(Locale.ROOT),
            uri.port,
            cleanPath(uri.path),
            null,
            null,
        ).toString()
        return normalized.trimEnd('/')
    }

    @JvmStatic
    fun isTrustedOrigin(serverUrl: String?, trustedOrigin: String?): Boolean {
        if (trustedOrigin.isNullOrBlank()) return false
        return runCatching {
            val server = URI(normalize(serverUrl))
            val trusted = URI(normalize(trustedOrigin))
            server.scheme.equals(trusted.scheme, ignoreCase = true) &&
                server.host.equals(trusted.host, ignoreCase = true) &&
                originPort(server) == originPort(trusted)
        }.getOrDefault(false)
    }

    @JvmStatic
    fun isPrivateLan(serverUrl: String): Boolean {
        val uri = URI(normalize(serverUrl))
        val host = uri.host.lowercase(Locale.ROOT)
        return host == "localhost" || host == "127.0.0.1" || host == "::1" ||
            host.startsWith("10.") || host.startsWith("192.168.") ||
            host.matches(Regex("172\\.(1[6-9]|2\\d|3[01])\\..*")) ||
            host.endsWith(".local")
    }

    private fun originPort(uri: URI): Int = when {
        uri.port >= 0 -> uri.port
        uri.scheme.equals("https", ignoreCase = true) -> 443
        else -> 80
    }

    private fun cleanPath(path: String?): String? = path
        ?.takeUnless { it.isBlank() || it == "/" }
        ?.trimEnd('/')
}
